### PR TITLE
kubernetes: cluster dump/restore instructions

### PR DESCRIPTION
**Pull request by colhom**
@robszumski @joshix still need to test this to make sure the instructions work verbatim, but otherwise ready to review.

\cc @philips

### CODE REVIEW

File: kubernetes/cluster-dump-restore.md [hunk 1]
```diff
@@ -0,0 +1,91 @@
...
+
+	As a general rule, if the *target cluster's* Kubernetes version is >= *source cluster's* Kubernets version, this will not be a problem.
+
+* **blank target slate** The *target cluster* to has no deployment state beyond a "stock" cluster. In the CoreOS case, the stock cluster state is everything in the `kube-system` namespace along with the service token in the default namespace.
+
+	If the *target cluster* has other pre-existing state, more precautions must be taken to ensure there are no undesired interactions between pre-existing resources on the *target cluster* and the dump coming from *source cluster*.
+
+	Name conflicts are easy to defect, but more subtle issues like services selecting unintended pre-existing pods is another more subtle situation that can arise. We will assume **blank target slate** for the remainder of this document.
+
+As of now, this is not entirely supported upstream. This `get --export` flag is supposed to do something like this, but leaves plenty of state that we do not consider to be portable in this case. (TODO: link to issue).
```
Comment (robszumski) [inline kubernetes/cluster-dump-restore.md]: make sure to fill this in

File: kubernetes/cluster-dump-restore.md [hunk 2]
```diff
@@ -0,0 +1,91 @@
...
+
+We'll explain step-by-step how the CoreOS infrastructure team migrates deployments between Kubernets clusters today.
```
Comment (robszumski) [inline kubernetes/cluster-dump-restore.md]: Kubernetes

File: kubernetes/cluster-dump-restore.md [hunk 3]
```diff
@@ -0,0 +1,91 @@
...
+```shell
+mkdir ./cluster-dump
+```
+
+
+Get a list all namespaces that are not `kube-system` or `default`. This represents the list of namespaces that we want to migrate:
+
+```shell
+kubectl get --export -o=json ns | \
+jq '.items[] |
```
Comment (robszumski) [inline kubernetes/cluster-dump-restore.md]: jq won't work if you're on CoreOS. Do we need to tell you to run this from a laptop? Should that be in a prereq section?

File: kubernetes/cluster-dump-restore.md [hunk 4]
```diff
@@ -0,0 +1,91 @@
...
+### Dump from the source cluster
+
+Create a directory to hold your dump files:
+
+```shell
+mkdir ./cluster-dump
+```
+
+
+Get a list all namespaces that are not `kube-system` or `default`. This represents the list of namespaces that we want to migrate:
```
Comment (robszumski) [inline kubernetes/cluster-dump-restore.md]: We should mention that we're writing these out to a file, for folks that might not be super familiar with more advanced bash usage. Maybe:

> **First,** get a list all namespaces that are not `kube-system` or `default` **and record them into a file on disk.**

File: kubernetes/cluster-dump-restore.md [hunk 5]
```diff
@@ -0,0 +1,91 @@
...
+            .status,
+            .spec.template.spec.securityContext,
+            .spec.template.spec.dnsPolicy,
+            .spec.template.spec.terminationGracePeriodSeconds,
+            .spec.template.spec.restartPolicy
+        )' >> "./cluster-dump/cluster-dump.json"
+done
+```
+
+Notice that pods and service tokens are explicitly omitted altogether, as they are inherently non-portable resources that are created and managed by other components. The general rule for what is portable across heterogenous clusters is is "services, controllers and secrets that aren't service tokens".
```
Comment (robszumski) [inline kubernetes/cluster-dump-restore.md]: Seems like this should be mentioned higher up in the page: "services, controllers and secrets that aren't service tokens"

Comment (robszumski) [inline kubernetes/cluster-dump-restore.md]: What should I be looking at to see that these two things are omitted? Or is it just that they aren't in the list?

File: kubernetes/cluster-dump-restore.md [hunk 6]
```diff
@@ -0,0 +1,91 @@
...
+        .metadata.uid,
+        .metadata.selfLink,
+        .metadata.resourceVersion,
+        .metadata.creationTimestamp,
+        .metadata.generation
+    )' > ./cluster-dump/ns.json
+```
+
+
+For each of these namespaces, dump all services, replicationcontrollers, secrets and daemonsets. Strip any non-portable fields from the objects. If you wish to migrate additional controller resource types (replicasets, deplopyments, etc), make sure to add them to the resource type list:
```
Comment (robszumski) [inline kubernetes/cluster-dump-restore.md]: deployments

**Comment by robszumski** [comment]
Two overall thoughts:

1. Is there a better word to use than "dump"?
2. Should this live in this repo or in the coreos-kubernetes repo? Either way it will end up at coreos.com/kubernetes, just wondering where we want it to live.

File: kubernetes/cluster-dump-restore.md (hunk 1 above)
Comment (colhom) [inline kubernetes/cluster-dump-restore.md]: I want to reference this document in the issue as a summary of what I think kubectl/api server should eventually do for you.

Here's [the issue](https://github.com/kubernetes/kubernetes/issues/21582). It's been radio silence for almost two months.

File: kubernetes/cluster-dump-restore.md (hunk 3 above)
Comment (colhom) [inline kubernetes/cluster-dump-restore.md]: CoreOS doesn't have kubectl either, so it's kind of implied you'll be doing this from your dev box.

On that topic, latest CoreOS actually ships with `jq`! Now if we just started shipping kubectl as well.

File: kubernetes/cluster-dump-restore.md (hunk 5 above)
Comment (colhom) [inline kubernetes/cluster-dump-restore.md]: For service tokens: `select(.type!="kubernetes.io/service-account-token")`

For pods, it's just that the `pod` resource type is omitted.

**Comment by colhom** [comment]
Agree that dump is lacking in the aesthetic appeal department. My rationale was honestly the term "database dump", tools like pg_dump, etc.

If not `dump`, I would have used `export`, but kubectl already has an `--export` option which does not really do what we need it to in this case.

File: kubernetes/cluster-dump-restore.md (hunk 4 above)
Comment (colhom) [inline kubernetes/cluster-dump-restore.md]: Agree

File: kubernetes/cluster-dump-restore.md [hunk 7]
```diff
@@ -0,0 +1,91 @@
...
+
+This document's primary purpose is to instruct how to migrate the deployment state from one Kubernetes cluster to another. The clusters may have different versions, pod/service network cidrs, number of nodes, etc.
+
+For the remainer of this document, the cluster that is being dumped will be referred to as the *source cluster*. The cluster that is being restored to will be called the *target cluster*. The goal is to migrate state from the *source cluster* to the *target cluster*.
+
+This migration process boils down to selectively omitting state that is not portable between clusters, as well as making some assumptions about the clusters we're dealing with. Those assumptions include:
+* **kube-dns add-on** This allows abstracting away the services' clusterIPs, which is inherentely non-portable across clusters with different serviceCIDRS.
+
+* **api object compatiblilty** Any api object type (eg: replicaset) that is dumped from the *source cluster* must be supported by the *target cluster*.
```
Comment (joshix) [inline kubernetes/cluster-dump-restore.md]: compatibility

File: kubernetes/cluster-dump-restore.md (hunk 6 above)
Comment (joshix) [inline kubernetes/cluster-dump-restore.md]: s/replicationcontrollers/replication controllers/

**Comment by colhom** [comment]
@robszumski nits addressed, I've also necro-bumped the GH issue.

File: kubernetes/cluster-dump-restore.md [hunk 8]
```diff
@@ -0,0 +1,93 @@
+## Migration of Kubernetes cluster deployment state
+
+_The deployment state of the Kubernetes cluster is stored in etcd. If you're concerned about backing up this information, you should look into backing up the etcd data directory for each etcd instance in your cluster. This can be done via an etcd-based backup strategy [example](https://github.com/coreos/etcd/tree/master/contrib/systemd/etcd2-backup-coreos), or via snapshotting the underlying block device that backs the etcd data directory. Backing up Kubernetes clusters is not the purpose of this document._
```
Comment (joshix) [inline kubernetes/cluster-dump-restore.md]: `s/etcd-based backup strategy [example]/[etcd-based backup strategy]/`

File: kubernetes/cluster-dump-restore.md [hunk 9]
```diff
@@ -0,0 +1,93 @@
...
+
+This document's primary purpose is to instruct how to migrate the deployment state from one Kubernetes cluster to another. The clusters may have different versions, pod/service network cidrs, number of nodes, etc.
```
Comment (joshix) [inline kubernetes/cluster-dump-restore.md]: s/instruct/show/

File: kubernetes/cluster-dump-restore.md [hunk 10]
```diff
@@ -0,0 +1,93 @@
...
+
+For the remainer of this document, the cluster that is being dumped will be referred to as the *source cluster*. The cluster that is being restored to will be called the *target cluster*. The goal is to migrate state from the *source cluster* to the *target cluster*.
```
Comment (joshix) [inline kubernetes/cluster-dump-restore.md]: s/remainer/remainder/

File: kubernetes/cluster-dump-restore.md [hunk 11]
```diff
@@ -0,0 +1,93 @@
...
+
+This migration process boils down to selectively omitting state that is not portable between clusters, as well as making some assumptions about the clusters we're dealing with. Those assumptions include:
+* **kube-dns add-on** This allows abstracting away the services' clusterIPs, which is inherentely non-portable across clusters with different serviceCIDRS.
+
+* **api object compatibilty** Any api object type (eg: replicaset) that is dumped from the *source cluster* must be supported by the *target cluster*.
```
Comment (danielfm) [inline kubernetes/cluster-dump-restore.md]: s/compatibilty/compatibility/

File: kubernetes/cluster-dump-restore.md [hunk 12]
```diff
@@ -0,0 +1,93 @@
...
+        .metadata.uid,
+        .metadata.selfLink,
+        .metadata.resourceVersion,
+        .metadata.creationTimestamp,
+        .metadata.generation
+    )' > ./cluster-dump/ns.json
+```
+
+
+For each of these namespaces, dump all services, replication controllers, secrets and daemonsets to a file on disk. Strip any non-portable fields from the objects. If you wish to migrate additional controller resource types (replicasets, deployments, etc), make sure to add them to the resource type list:
```
Comment (cgag) [inline kubernetes/cluster-dump-restore.md]: Unless there's a good reason you wouldn't want to restore these things, we should just add them to the list in the code and in the opening sentence and delete the "if you wish..." stuff.

File: kubernetes/cluster-dump-restore.md [hunk 13]
```diff
@@ -0,0 +1,93 @@
...
+            .status,
+            .spec.template.spec.securityContext,
+            .spec.template.spec.dnsPolicy,
+            .spec.template.spec.terminationGracePeriodSeconds,
+            .spec.template.spec.restartPolicy
+        )' >> "./cluster-dump/cluster-dump.json"
+done
+```
+
+Notice that pods and service tokens are explicitly omitted altogether, as they are inherently non-portable resources that are created and managed by other components. The general rule for what is portable across heterogenous clusters is is "services (resolved via cluster DNS), controllers and secrets that aren't service tokens".
```
Comment (cgag) [inline kubernetes/cluster-dump-restore.md]: I'd rather say "replication controllers" specifically, or something like "controllers (replication controllers, replicasets, deployments)", or ideally just "[...], replication controllers, replicasets, deployements, [...]".

**Comment by philips** [comment]
How do we get this merged?

**Comment by colhom** [comment]
@philips I'll hit these zombie nits and we'll do it in about 15 min

**Comment by colhom** [comment]
@robszumski @philips this should be ready to go. Give word and I'll squash.

**Comment by robszumski** [comment]
LGTM

**Pull request merged**